packages/modules/payment/src/providers/payment-medusa/services/medusa-payments.ts

```diff
@@ -0,0 +1,634 @@
+      capture_method: extra?.capture_method as "automatic" | "manual",
+      setup_future_usage: extra?.setup_future_usage,
+      payment_method_types: extra?.payment_method_types ?? "",
+      payment_method_data: extra?.payment_method_data,
+      payment_method_options: extra?.payment_method_options,
+      automatic_payment_methods: extra?.automatic_payment_methods,
+      off_session: extra?.off_session,
+      confirm: extra?.confirm,
+      payment_method: extra?.payment_method,
+      return_url: extra?.return_url,
+      shared_payment_token: extra?.shared_payment_token,
+    }
+
+    return res
+  }
+
+  async executeWithRetry<T>(
+    apiCall: () => Promise<T>,
+    maxRetries: number = 3,
+    baseDelay: number = 1000,
+    currentAttempt: number = 1
+  ): Promise<T> {
+    try {
+      return await apiCall()
+    } catch (error) {
+      const handledError = { retry: false, data: error }
+
+      if (!handledError.retry) {
+        // If retry is false, we know data exists per the type definition
+        return handledError.data
+      }
+
+      if (handledError.retry && currentAttempt <= maxRetries) {
+        // Logic for retrying
+        const delay =
+          baseDelay *
+          Math.pow(2, currentAttempt - 1) *
+          (0.5 + Math.random() * 0.5)
+        await setTimeout(delay)
+        return this.executeWithRetry(
+          apiCall,
+          maxRetries,
+          baseDelay,
+          currentAttempt + 1
+        )
+      }
+
+      // Retries are exhausted
+      throw new Error(
+        `An error occurred in InitiatePayment during creation of medusa payments payment intent: ${error.message}. ${error.stack}`
+      )
+    }
+  }
+
+  async getPaymentStatus(
+    input: GetPaymentStatusInput
+  ): Promise<GetPaymentStatusOutput> {
+    const id = input?.data?.id as string
+    if (!id) {
+      throw new Error(
+        "No payment intent ID provided while getting payment status"
+      )
+    }
+
+    const payment = await this.retrievePayment({ data: { id } })
+
+    return {
+      status: payment.data?.status as PaymentSessionStatus,
+      data: payment.data,
+    }
+  }
+
+  async initiatePayment({
+    currency_code,
+    amount,
+    data,
+    context,
+  }: InitiatePaymentInput): Promise<InitiatePaymentOutput> {
+    const additionalParameters = this.normalizePaymentParameters(
+      data as Record<string, string>
+    )
+
+    const intentRequest = {
+      amount: getSmallestUnit(amount, currency_code),
+      currency: currency_code,
+      metadata: {
+        session_id: data?.session_id as string,
+      },
+      account_holder_id: context?.account_holder?.data?.id as
+        | string
+        | undefined,
+      idempotency_key: context?.idempotency_key,
+      ...additionalParameters,
+    }
+
+    const payment = await this.executeWithRetry(() => {
+      return this.request<{ payment: any }>("/payments", {
+        method: "POST",
+        body: JSON.stringify(intentRequest),
```

Comment:
nit(non-blocking): we could bake the stringifying into the `request`, so we can avoid having/remember to do this